### DISCUSSION TITLE
Why is my Webfinger not working with Mastodon?

### DISCUSSION THREAD

**Discussion by pessi-v** [mastodon/mastodon#30712] [General]
I'm trying to make a simple implementation of AP for a project, and setting up Webfinger is kind of the first step. It's already working with some other clients, but not Mastodon.

Could you take a look and tell me what's missing?

Note: the server might need 15-30s to get the app running after a request
[https://climatenews-xyz.fly.dev/.well-known/webfinger?resource=acct:aggregator@climatenews-xyz.fly.dev](https://climatenews-xyz.fly.dev/.well-known/webfinger?resource=acct:aggregator@climatenews-xyz.fly.dev)

Example of a Fediverse client where Webfinger does work:
https://community.nodebb.org/user/aggregator@climatenews-xyz.fly.dev

**Comment by 0xWDG** [comment]
My webfinger doesnt work as well, https://wesleydegroot.nl/.well-known/webfinger?resource=acct:0xWDG@wesleydegroot.nl
I'd like to have a debugger/tester for such things, I have no idea why it isn't working.

**Reply by whalygood** [reply]
How did you end up getting this working @0xWDG ? 

I can see that your handle 0xWED@wesleydegroot.nl is now searchable on Mastodon instances such as mastodon.social.
I am having the same as @pessi-v where community.nodebb.org is able to pull data from my webfinger and actor page (name, avatar, bio) but Mastodon returns a "Could not find anything for these search terms" when I search for username@domain.com.

The NodeBB client is even able to pull changes, such as if I update my bio, so it's clearly something funky with Mastodon. 

This is from my NGINX logs which shows Mastodon makes a request to my webfinger, then follows the "rel": "self" to find my actor object successfully. Both requests from Mastodon return a 200 status code, and I don't have any other logs to understand why Mastodon is not accepting something in my JSON-LD.

```
168.119.49.146 - - [18/Aug/2024:21:22:39 +0000] "GET /.well-known/webfinger?resource=acct:test1234@example.com HTTP/1.1" 200 289 "-" "Mastodon/4.3.0-nightly.2024-08-12 (http.rb/5.2.0; +https://mastodon.social/)"
168.119.49.146 - - [18/Aug/2024:21:22:39 +0000] "GET /activitypub/test1234 HTTP/1.1" 200 1848 "-" "Mastodon/4.3.0-nightly.2024-08-12 (http.rb/5.2.0; +https://mastodon.social/)"
```

**Did you have to use a particular key type for publicKeyPem**? - because that looks like the only thing which might be different between my implementation and what I see at https://wesleydegroot.nl/.well-known/webfinger and https://wesleydegroot.nl/0xWDG.

Any insight you could share would be much appreciated!

**Reply by whalygood** [reply]
This is my webfinger

```
{
    "subject": "acct:test1234@example.com",
    "aliases": [
        "https://example.com/test1234"
    ],
    "links": [
        {
            "rel": "self",
            "type": "application/activity+json",
            "href": "https://example.com/activitypub/test1234"
        },
        {
            "rel": "http://webfinger.net/rel/profile-page",
            "type": "text/html",
            "href": "https://example.com/test1234"
        }
    ]
}
```

And this is my actor object

```
{
    "@context": "https://www.w3.org/ns/activitystreams",
    "id": "https://example.com/test1234",
    "type": "Person",
    "following": "https://example.com/test1234/following",
    "followers": "https://example.com/test1234/followers",
    "inbox": "https://example.com/activitypub/inbox",
    "outbox": "https://example.com/activitypub/outbox",
    "preferredUsername": "test1234",
    "name": "test1234",
    "summary": "Sed tincidunt ac urna non commodo. Nunc magna turpis, tempor id erat non, sagittis dapibus nisi. Aliquam malesuada finibus porttitor. Suspendisse vehicula tincidunt nisi nec dapibus. Donec facilisis mattis augue, tristique ultrices diam semper ut. In faucibus rutrum nisl at rhoncus. Quisque et volutpat justo. Fusce aliquam neque eu viverra tempus. Suspendisse non molestie dui. Curabitur eu dui porttitor, dignissim tortor sed, eleifend elit. Pellentesque vel erat lorem.",
    "url": "https://example.com/test1234",
    "manuallyApprovesFollowers": false,
    "discoverable": true,
    "memorial": false,
    "published": "1970-01-01T00:00:00Z",
    "icon": {
        "type": "Image",
        "mediaType": "image/png",
        "url": "https://api.example.com:443/images/artist-avatars/5ff1ecb9-f320-4a1f-8dcc-bb41a50cbf61-x600.jpg"
    },
    "image": {
        "type": "Image",
        "mediaType": "image/png",
        "url": "https://api.example.com:443/images/artist-banners/5aa11d1f-a4b6-40af-909b-65dbc50f7148-x2500.jpg"
    },
    "publicKey": {
        "id": "https://example.com/test1234#main-key",
        "owner": "https://example.com/test1234",
        "publicKeyPem": "-----BEGIN PUBLIC KEY-----\nMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAoGAmhuIAvtrli2s8D3jF\n/UrgbYi4hm94dDBEafJ5GF2KUJAPtchaQerKQQaTfIefFPsiG2QGvnv+QHQO9+dU\nijcS0gvlILkpMJhrgRi/qZovXwpNBbzTkX8odP8tuajhhDqGNFKmptNn2kUIIXTA\ndQ8Se0VgUSX4GV00V7BfpGv9H9LExx8QzGjKB25tTzkVN9/KwuEWjYZy0yRlhuT8\nX4LMG2hh9HHteFsgVRkbyuJgjidtdlX6KwPRQwN9HK0Lv5V7QCY6nktZSCReydlb\n2iMMRpH5VxHXKDdWvJHUv50pzwGWrsqFY/+mLYVnT+C3Y8DNebG2UfyyNk9YXaET\nqQIDAQAB\n-----END PUBLIC KEY-----\n"
    }
}
```

I can see that even aggregator@climatenews-xyz.fly.dev (@pessi-v 's ActivityPub account) can be discovered now on Mastodon instances. So I am wondering what is the discrepancy here.

**Reply by whalygood** [reply]
Got it working (discoverable on Mastodon)! 
Had to change these two things, if it's helps anyone else who is trying to accomplish this:

1. In the actor object,
`"id": "https://example.com/test1234",` should be `"id": "https://example.com/webfinger/test1234",`
since my actor object is located at /webfinger/test1234 and not at /test1234.
2. Also I had to set up an outbox, even if I am not providing any activities to ActivityPub instances through it yet. Since my implementation is multi-user, I made sure that my Node service that generates the JSON generates a different outbox for each user. For example, the user "test1234" would have an outbox at /activitypub/test1234/outbox. 

For what it's worth, for troubleshooting, I also switched off Cloudflare's reverse proxy for the time being (since it can apparently sometimes prevent certain ActivityPub instances from federating properly), but that wasn't related in this case.

**Reply by 0xWDG** [reply]
My issue was that I was returning an content-type of `application/json` instead of `application/activity+json`.
The account is discoverable on Mastodon right now, but it fails to show any posts, I still need to fix that.

I used:
- https://activitypub.academy/activitypub_explorer (this did not show me any warnings of the incorrect content-type).
- https://socialweb.coop/activitypub/actor/tester/
